book/src/rust/rust_style_guide.md

```diff
@@ -0,0 +1,660 @@
+extern crate crate_name;
+
+use module::submodule::entry;
+
+#[cfg(feature = "...")]
+use module::feature_gated_submodule;
+
+mod another_module;
+mod module;
+mod the_third_module;
+
+#[cfg(feature = "...")]
+mod feature_gated_module;
+
+pub mod public_module;
+
+#[cfg(feature = "...")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "...")))]
+pub mod feature_gated_public_module;
+
+pub use reexported_objects;
+
+#[cfg(feature = "...")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "...")))]
+pub use feature_gated_reexported_objects;
+```
+
+Ensure that there are new lines between each category. Be sure that if some imports or exports are feature-gated, that they are sorted by feature alphabetically. If there is a feature gated import that requires importing multiple objects, use the following pattern:
+
+```rust
+#[cfg(feature = "...")]
+use {
+    thing1, thing2, thing3, thing4,
+};
+```
+
+**NOTE**: All imports should occur at the top of any module, and a newline should be added between the last import and the first declared object.
+
+#### Let binding
+
+
+##### Always separately bind [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) guards
+
+
+>RAII stands for "Resource Acquisition is Initialisation". The essence of the pattern is that resource initialisation is done in the constructor of an object and finalisation in the destructor. This pattern is extended in Rust by using a RAII object as a guard of some resource and relying on the type system to ensure that access is always mediated by the guard object.
+
+Prefer
+
+```rust
+fn use_mutex(m: sync::mutex::Mutex<int>) {
+    let guard = m.lock();
+    do_work(guard);
+    drop(guard); // unlock the lock
+    // do other work
+}
+```
+over
+
+```rust
+fn use_mutex(m: sync::mutex::Mutex<int>) {
+    do_work(m.lock());
+    // do other work
+}
+```
+
+##### Prefer conditional expressions to deferred initialization.
+
+Prefer
+
+```rust
+let foo = match bar {
+    Baz  => 0,
+    Quux => 1
+};
+```
+
+Over
+
+```rust
+let foo;
+match bar {
+    Baz  => {
+        foo = 0;
+    }
+    Quux => {
+        foo = 1;
+    }
+}
+```
+
+unless the conditions for initialization are too complex to fit into a simple conditional expression.
+
+##### Use type annotations for clarification; prefer explicit generics when inference fails.
+
+Prefer
+
+```rust
+s.iter().map(|x| x * 2)
+        .collect::<Vec<_>>()
+```
+
+Over
+
+```rust
+let v: Vec<_> = s.iter().map(|x| x * 2)
+                        .collect();
+```
+
+When the type of a value might be unclear to the reader of the code, consider explicitly annotating it in a let.
+
+On the other hand, when the type is unclear to the compiler, prefer to specify the type by explicit generics instantiation, which is usually more clear.
+
+##### Prefer immutable bindings.
+
+Use mut bindings to signal the span during which a value is mutated:
+
+```rust
+let mut v = Vec::new();
+// push things onto v
+let v = v;
+// use v immutably henceforth
+```
```

Comment:
This seems more confusing than helpful. If I saw `let v = v;`, I'd assume it was a mistake rather than intentionally removing `mut`